plugins/inventory/vmware_vm_inventory.py

```diff
@@ -395,211 +485,255 @@ def parse(self, inventory, loader, path, cache=True):
+                    c_group = p_group
+                    parents.pop()
+
+        # Hostvars formats manipulation
+        host_properties = host_properties if with_nested_properties else to_flatten_dict(host_properties)
+
+        if self.get_option('with_sanitized_property_name'):
+            host_properties = camel_dict_to_snake_dict(host_properties)
+
+        can_sanitize = self.get_option('with_sanitized_property_name')
+        for k, v in host_properties.items():
+            k = self._sanitize_group_name(k) if can_sanitize else k
+            self.inventory.set_variable(host, k, v)
+
+
+def parse_vim_property(vim_prop, key=""):
+    # For '--yaml' sake!
+    #   Unexpected Exception, this is probably a bug: ('cannot represent an object', <data>
+    prop_type = type(vim_prop).__name__
+    if prop_type.startswith("vim") or prop_type.startswith("vmodl"):
+        if isinstance(vim_prop, DataObject):
+            r = {}
+            for prop in vim_prop._GetPropertyList():
+                if prop.name not in ['dynamicProperty', 'dynamicType', 'managedObjectType']:
+                    sub_prop = getattr(vim_prop, prop.name)
+                    r[prop.name] = parse_vim_property(sub_prop, prop.name)
+            return r
+
+        elif isinstance(vim_prop, list):
+            r = []
+            for prop in vim_prop:
+                r.append(parse_vim_property(prop))
+            return r
+        return vim_prop.__str__()
+
+    elif prop_type == "datetime":
+        return Iso8601.ISO8601Format(vim_prop)
+
+    elif prop_type == "long":
+        return int(vim_prop)
+    elif prop_type == "long[]":
+        return [int(x) for x in vim_prop]
+
+    elif isinstance(vim_prop, list):
+        return [parse_vim_property(x) for x in vim_prop]
+
+    elif prop_type in ['bool', 'int', 'NoneType']:
+        return vim_prop
+
+    return to_text(vim_prop)
+
+
+def to_nested_dict(vm_properties):
+    host_properties = {}
+
+    for vm_prop_name, vm_prop_val in vm_properties.items():
+        prop_parents = reversed(vm_prop_name.split("."))
+        prop_dict = parse_vim_property(vm_prop_val, vm_prop_name)
+
+        for k in prop_parents:
+            prop_dict = {k: prop_dict}
+        host_properties = dict_merge(host_properties, prop_dict)
+
+    return host_properties
+
+
+def to_flatten_dict(d, parent_key='', sep='.'):
+    items = []
+    for k, v in d.items():
+        new_key = parent_key + sep + k if parent_key else k
+        if v and isinstance(v, dict):
+            items.extend(to_flatten_dict(v, new_key, sep=sep).items())
+        else:
+            items.append((new_key, v))
+    return dict(items)
+
+
+# Patch AnsibleDumper
+# This for 'ansible-test' with '--debug'
+# TODO: Move it to 'ansible.parsing.yaml.dumper'
+from ansible.vars.manager import VarsWithSources
+from ansible.parsing.yaml.dumper import AnsibleDumper, represent_hostvars
+
+AnsibleDumper.add_representer(VarsWithSources, represent_hostvars)
```

Comment:
The critical difference I see here is that, while this is already used as a representer for other types, `VarsWithSources` was a recent merge: https://github.com/ansible/ansible/pull/66102

Given that, I think there is a decent shot of getting this in Ansible base, but the issue should be raised sooner rather than later, and it would be good to have a reproducer for the issue it addresses.

Reply:
I would love to do a PR for it, if there is a no objection.

Reply:
Submitted a PR https://github.com/ansible/ansible/pull/68525.

Because this is a general bug in all inventory types with yaml and debug mode enabled, I'll remove this part.